internal/ui/esbuild.go

```diff
@@ -0,0 +1,173 @@
+	Splitting:   true,
+	Loader: map[string]api.Loader{
+		".scss": api.LoaderCSS, // Treat SCSS files as CSS
+	},
+}
+
+// getOptions updates the build options based on the dev flag
+// Dev mode uses inline source maps, and has a debug log level
+// Production mode minifies the output to boost performance, and logs only errors
+func getOptions(dev bool, outputDir *string) api.BuildOptions {
+	opts := buildOptions
+	if dev {
+		opts.Sourcemap = api.SourceMapInline
+		opts.LogLevel = api.LogLevelDebug
+	} else {
+		opts.LogLevel = api.LogLevelError
+		opts.MinifyWhitespace = true
+		opts.MinifyIdentifiers = true
+		opts.MinifySyntax = true
+	}
+	// This is done to enable testing without overwriting current build
+	if outputDir != nil {
+		opts.Outdir = *outputDir
+	}
+	return opts
+}
+
+// Build builds the UI with esbuild
+// The entry point is src/index.tsx and the output is public/dist
+// Scss files are treated as css
+func Build(dev bool, outputDir *string) error {
+	result := api.Build(getOptions(dev, outputDir))
+	if len(result.Errors) > 0 {
+		return fmt.Errorf("failed to build UI: %v", parseMessages(result.Errors))
+	}
+	if err := createHtml(); err != nil {
+		return fmt.Errorf("failed to create index.html: %v", err)
+	}
+	return nil
+}
+
+// createIndexTemplate creates the index template for the UI
+// It injects script references into the index template
+// The index template is located in public/index.tmpl.html
+// The script references are located in public/dist
+// The index template is written to public/assets/index.html
+func createHtml() error {
+	indexTmpl := fmt.Sprintf("%s/index.tmpl.html", env.PublicDir())
+	f, err := os.ReadFile(indexTmpl)
+	if err != nil {
+		return fmt.Errorf("failed to stat index template: %v", err)
+	}
+	lines := removeComments(strings.Split(string(f), "\n"))
+
+	indexToInject := findClosingHeadTag(lines)
+	if indexToInject == -1 {
+		return errors.New("failed to find </head> in index template")
+	}
+
+	links, err := getLinks()
+	if err != nil {
+		return fmt.Errorf("failed to get script refs: %v", err)
+	}
+
+	// Inject script references into index template and write to public/assets/index.html
+	lines = append(lines[:indexToInject], append(links, lines[indexToInject:]...)...)
+	if err := os.WriteFile(fmt.Sprintf("%s/assets/index.html", env.PublicDir()), []byte(strings.Join(lines, "\n")), 0644); err != nil {
+		return fmt.Errorf("failed to write index.html: %v", err)
+	}
+	return nil
+}
+
+// findClosingHeadTag attempts to find the closing head tag in the index template
+// Returns index of the line before the closing head tag
+// The closing head tag is defined as </head>
+func findClosingHeadTag(lines []string) int {
+	for i, line := range lines {
+		if strings.Contains(line, "</head>") {
+			return i
+		}
+	}
+	return -1
+}
+
+// getLinks gets the script and css references from the dist directory
+// Support extensions are .css and .js
+func getLinks() ([]string, error) {
```

Comment:
Note that the `.js` and `.css` files are returned from `api.Build(...)`, so I'm not sure we need this logic.

```go
result := api.Build(getOptions(dev, outputDir))
...
for _, file := range result.OutputFiles {
    // file.Path includes the absolute path to the generated file
}
```

You could also consider using templates to inject the file links into the `index.html` file

```
...
{{range .CSSFiles}}<link rel="stylesheet" href="/static/{{.}}">{{end}}
{{range .JSFiles}}<script type="module" src="/static/{{.}}" defer></script>{{end}}
...
```